src/librustc_codegen_llvm/lib.rs

```diff
@@ -270,6 +272,47 @@ impl CodegenBackend for LlvmCodegenBackend {
         )
     }
 
+    fn link(
+        &self,
+        sess: &Session,
+        input: &Input,
+        outputs: &OutputFilenames,
+    ) -> Result<(), ErrorReported> {
+        sess.compile_status()?;
+        let rlink_file = match input {
+            Input::File(file) => file,
+            _ => bug!("rlink must be a file"),
+        };
+        let rlink_data = fs::read_to_string(rlink_file).map_err(|err| {
+            sess.fatal(&format!("failed to read rlink file: {}", err));
+        })?;
+        let codegen_results: CodegenResults = json::decode(&rlink_data).map_err(|err| {
+            sess.fatal(&format!("failed to decode rlink: {}", err));
+        })?;
+
+        // Run the linker on any artifacts that resulted from the LLVM run.
+        // This should produce either a finished executable or library.
+        sess.time("link_crate", || {
+            use crate::back::archive::LlvmArchiveBuilder;
+            use rustc_codegen_ssa::back::link::link_binary;
+
+            let target_cpu = crate::llvm_util::target_cpu(sess);
+            link_binary::<LlvmArchiveBuilder<'_>>(
+                sess,
+                &codegen_results,
+                outputs,
+                &codegen_results.crate_name.as_str(),
+                target_cpu,
+            );
+        });
+
+        // Now that we won't touch anything in the incremental compilation directory
+        // any more, we can finalize it (which involves renaming it)
+        rustc_incremental::finalize_session_directory(sess, codegen_results.crate_hash);
```

Comment:
This is already performed when creating the `.rlink` file. I think it can be moved before linking for `join_codegen_and_link` too.

Reply:
Will do once #68601 is merged.